File: ContractionTimer/src/android/java/com/ianhanniballake/contractiontimer/ui/ViewActivity.java
package com.ianhanniballake.contractiontimer.ui;

import android.content.ContentUris;
import android.content.Context;
import android.content.SharedPreferences;
import android.content.pm.ActivityInfo;
import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentStatePagerAdapter;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.CursorLoader;
import android.support.v4.content.Loader;
import android.support.v4.view.ViewPager;
import android.support.v4.widget.CursorAdapter;
import android.support.v7.app.ActionBarActivity;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.google.android.gms.tagmanager.DataLayer;
import com.ianhanniballake.contractiontimer.BuildConfig;
import com.ianhanniballake.contractiontimer.R;
import com.ianhanniballake.contractiontimer.provider.ContractionContract;
import com.ianhanniballake.contractiontimer.provider.ContractionContract.Contractions;
import com.ianhanniballake.contractiontimer.tagmanager.GtmManager;

import org.acra.ACRA;

/**
 * Stand alone activity used to view the details of an individual contraction
 */
public class ViewActivity extends ActionBarActivity implements LoaderManager.LoaderCallbacks<Cursor>,
        ViewPager.OnPageChangeListener {
    private final static String TAG = ViewActivity.class.getSimpleName();
    /**
     * Adapter for all contractions
     */
    CursorAdapter adapter = null;
    /**
     * Currently shown page
     */
    int currentPosition = -1;
    /**
     * Pager Adapter to manage view contraction pages
     */
    private ViewFragmentPagerAdapter pagerAdapter;

    @Override
    public void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_view);
        if (findViewById(R.id.pager) == null) {
            // A null pager means we no longer need this activity
            finish();
            return;
        }
        adapter = new CursorAdapter(this, null, 0) {
            @Override
            public void bindView(final View view, final Context context, final Cursor cursor) {
                // Nothing to do
            }

            @Override
            public View newView(final Context context, final Cursor cursor, final ViewGroup parent) {
                return null;
            }
        };
        final ViewPager pager = (ViewPager) findViewById(R.id.pager);
        pager.setOnPageChangeListener(this);
        pagerAdapter = new ViewFragmentPagerAdapter(getSupportFragmentManager());
        getSupportLoaderManager().initLoader(0, null, this);
    }

    @Override
    public Loader<Cursor> onCreateLoader(final int id, final Bundle args) {
        return new CursorLoader(this, ContractionContract.Contractions.CONTENT_URI, null, null, null,
                Contractions.COLUMN_NAME_START_TIME);
    }

    @Override
    public boolean onCreateOptionsMenu(final Menu menu) {
        getMenuInflater().inflate(R.menu.activity_view, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public void onLoaderReset(final Loader<Cursor> loader) {
        adapter.swapCursor(null);
        pagerAdapter.notifyDataSetChanged();
    }

    @Override
    public void onLoadFinished(final Loader<Cursor> loader, final Cursor data) {
        adapter.swapCursor(data);
        pagerAdapter.notifyDataSetChanged();
        // A null content uri means we should cancel out as we are in an indeterminate state
        final Uri contentUri = getIntent() == null ? null : getIntent().getData();
        if (contentUri == null) {
            finish();
            return;
        }
        final long contractionId;
        try {
            contractionId = ContentUris.parseId(contentUri);
        } catch (final NumberFormatException e) {
            if (BuildConfig.DEBUG)
                Log.e(TAG, "NumberFormatException in onLoadFinished", e);
            else {
                GtmManager.getInstance(this).pushException(e);
                ACRA.getErrorReporter().handleSilentException(e);
            }
            finish();
            return;
        }
        final int count = adapter.getCount();
        for (int position = 0; position < count; position++) {
            final long id = adapter.getItemId(position);
            if (id == contractionId) {
                currentPosition = position;
                break;
            }
        }
        if (currentPosition == -1)
            finish();
        final ViewPager pager = (ViewPager) findViewById(R.id.pager);
        pager.setAdapter(pagerAdapter);
        pager.setCurrentItem(currentPosition, false);
    }

    @Override
    public boolean onOptionsItemSelected(final MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            if (BuildConfig.DEBUG)
                Log.d(TAG, "View selected home");
            GtmManager.getInstance(this).pushEvent("Home");
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onPageScrolled(final int position, final float positionOffset, final int positionOffsetPixels) {
        // Nothing to do
    }

    @Override
    public void onPageScrollStateChanged(final int state) {
        // Nothing to do
    }

    @Override
    public void onPageSelected(final int position) {
        if (BuildConfig.DEBUG)
            Log.d(TAG, "Swapped to " + position);
        if (currentPosition != -1) {
            final String direction = position > currentPosition ? "Next" : "Previous";
            GtmManager.getInstance(this).pushEvent("Scroll",
                    DataLayer.mapOf("direction", direction, "position", position));
        }
        currentPosition = position;
        final long newContractionId = adapter.getItemId(position);
        getIntent().setData(
                ContentUris.withAppendedId(ContractionContract.Contractions.CONTENT_ID_URI_BASE, newContractionId));
    }

    @Override
    protected void onResume() {
        super.onResume();
        final SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this);
        final boolean isLockPortrait = preferences.getBoolean(Preferences.LOCK_PORTRAIT_PREFERENCE_KEY, getResources()
                .getBoolean(R.bool.pref_settings_lock_portrait_default));
        if (BuildConfig.DEBUG)
            Log.d(TAG, "Lock Portrait: " + isLockPortrait);
        if (isLockPortrait)
            setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
        else
            setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_SENSOR);
    }

    @Override
    protected void onStart() {
        super.onStart();
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        GtmManager.getInstance(this).pushOpenScreen("View");
        if (getIntent().hasExtra(MainActivity.LAUNCHED_FROM_WIDGET_EXTRA)) {
            final String widgetIdentifier = getIntent().getStringExtra(MainActivity.LAUNCHED_FROM_WIDGET_EXTRA);
            if (BuildConfig.DEBUG)
                Log.d(TAG, "Launched from " + widgetIdentifier);
            GtmManager.getInstance(this).pushEvent("LaunchView", DataLayer.mapOf("widget", widgetIdentifier,
                    "type", DataLayer.OBJECT_NOT_PRESENT));
            getIntent().removeExtra(MainActivity.LAUNCHED_FROM_WIDGET_EXTRA);
        }
    }

    /**
     * Creates ViewFragments as necessary
     */
    private class ViewFragmentPagerAdapter extends FragmentStatePagerAdapter {
        /**
         * Creates a new ViewFragmentPagerAdapter
         *
         * @param fm FragmentManager used to manage fragments
         */
        public ViewFragmentPagerAdapter(final FragmentManager fm) {
            super(fm);
        }

        @Override
        public int getCount() {
            return adapter == null ? 0 : adapter.getCount();
        }

        @Override
        public Fragment getItem(final int position) {
            return ViewFragment.createInstance(adapter.getItemId(position));
        }

        @Override
        public CharSequence getPageTitle(final int position) {
            if (position + 1 == currentPosition)
                return getText(R.string.detail_previous_page);
            else if (position - 1 == currentPosition)
                return getText(R.string.detail_next_page);
            else
                return null;
        }
    }
}

File: mobile/src/main/java/com/ianhanniballake/contractiontimer/ui/ViewActivity.kt
package com.ianhanniballake.contractiontimer.ui

import android.content.ContentUris
import android.content.Context
import android.content.pm.ActivityInfo
import android.database.Cursor
import android.os.Bundle
import android.preference.PreferenceManager
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.view.ViewCompat
import androidx.cursoradapter.widget.CursorAdapter
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentStatePagerAdapter
import androidx.loader.app.LoaderManager
import androidx.loader.content.CursorLoader
import androidx.loader.content.Loader
import androidx.viewpager.widget.ViewPager
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.ianhanniballake.contractiontimer.BuildConfig
import com.ianhanniballake.contractiontimer.R
import com.ianhanniballake.contractiontimer.provider.ContractionContract.Contractions

/**
 * Stand alone activity used to view the details of an individual contraction
 */
class ViewActivity : AppCompatActivity(),
        LoaderManager.LoaderCallbacks<Cursor>,
        ViewPager.OnPageChangeListener {
    companion object {
        private const val TAG = "ViewActivity"
    }

    /**
     * Adapter for all contractions
     */
    internal lateinit var adapter: CursorAdapter
    /**
     * Currently shown page
     */
    internal var currentPosition = -1
    /**
     * Pager Adapter to manage view contraction pages
     */
    private lateinit var pagerAdapter: ViewFragmentPagerAdapter

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_view)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        if (findViewById<View>(R.id.pager) == null) {
            // A null pager means we no longer need this activity
            finish()
            return
        }
        adapter = object : CursorAdapter(this, null, 0) {
            override fun bindView(view: View, context: Context, cursor: Cursor) {
                // Nothing to do
            }

            override fun newView(context: Context, cursor: Cursor, parent: ViewGroup): View? {
                return null
            }
        }
        val pager = findViewById<ViewPager>(R.id.pager)
        pager.addOnPageChangeListener(this)
        pagerAdapter = ViewFragmentPagerAdapter(supportFragmentManager)
        supportLoaderManager.initLoader(0, null, this)
    }

    override fun onCreateLoader(id: Int, args: Bundle?): Loader<Cursor> {
        return CursorLoader(
            this, Contractions.CONTENT_URI, null,
            null, null, Contractions.COLUMN_NAME_START_TIME
        )
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.activity_view, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onLoaderReset(loader: Loader<Cursor>) {
        adapter.swapCursor(null)
        pagerAdapter.notifyDataSetChanged()
    }

    override fun onLoadFinished(loader: Loader<Cursor>, data: Cursor) {
        adapter.swapCursor(data)
        pagerAdapter.notifyDataSetChanged()
        // A null content uri means we should cancel out as we are in an indeterminate state
        val contentUri = intent?.data
        if (contentUri == null) {
            finish()
            return
        }
        val contractionId: Long
        try {
            contractionId = ContentUris.parseId(contentUri)
        } catch (e: NumberFormatException) {
            if (BuildConfig.DEBUG)
                Log.e(TAG, "NumberFormatException in onLoadFinished", e)
            else {
                FirebaseCrashlytics.getInstance().recordException(e)
            }
            finish()
            return
        }

        val count = adapter.count
        for (position in 0 until count) {
            val id = adapter.getItemId(position)
            if (id == contractionId) {
                currentPosition = position
                break
            }
        }
        if (currentPosition == -1)
            finish()
        val pager = findViewById<ViewPager>(R.id.pager)
        pager.adapter = pagerAdapter
        pager.setCurrentItem(currentPosition, false)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            if (BuildConfig.DEBUG)
                Log.d(TAG, "View selected home")
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
        // Nothing to do
    }

    override fun onPageScrollStateChanged(state: Int) {
        // Nothing to do
    }

    override fun onPageSelected(position: Int) {
        if (BuildConfig.DEBUG)
            Log.d(TAG, "Swapped to $position")
        currentPosition = position
        val newContractionId = adapter.getItemId(position)
        intent.data = ContentUris.withAppendedId(
            Contractions.CONTENT_ID_URI_BASE, newContractionId
        )
    }

    override fun onResume() {
        super.onResume()
        val preferences = PreferenceManager.getDefaultSharedPreferences(this)
        val isLockPortrait = preferences.getBoolean(Preferences.LOCK_PORTRAIT_PREFERENCE_KEY,
                resources.getBoolean(R.bool.pref_lock_portrait_default))
        if (BuildConfig.DEBUG)
            Log.d(TAG, "Lock Portrait: $isLockPortrait")
        requestedOrientation = if (isLockPortrait)
            ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
        else
            ActivityInfo.SCREEN_ORIENTATION_SENSOR
    }

    override fun onStart() {
        super.onStart()
        supportActionBar?.apply {
            setDisplayHomeAsUpEnabled(true)
            elevation = 0f
        }
        ViewCompat.setElevation(findViewById(R.id.pager_title_strip),
                resources.getDimension(R.dimen.action_bar_elevation))
        if (intent.hasExtra(MainActivity.LAUNCHED_FROM_WIDGET_EXTRA)) {
            val widgetIdentifier = intent.getStringExtra(MainActivity.LAUNCHED_FROM_WIDGET_EXTRA)
            if (BuildConfig.DEBUG)
                Log.d(TAG, "Launched from $widgetIdentifier")
            FirebaseAnalytics.getInstance(this).logEvent("${widgetIdentifier}_view_launch", null)
            intent.removeExtra(MainActivity.LAUNCHED_FROM_WIDGET_EXTRA)
        }
    }

    /**
     * Creates ViewFragments as necessary
     */
    private inner class ViewFragmentPagerAdapter(
            fm: FragmentManager
    ) : FragmentStatePagerAdapter(fm) {

        override fun getCount(): Int {
            return adapter.count
        }

        override fun getItem(position: Int): Fragment {
            return ViewFragment.createInstance(adapter.getItemId(position))
        }

        override fun getPageTitle(position: Int): CharSequence? {
            return when (currentPosition) {
                position + 1 -> getText(R.string.detail_previous_page)
                position - 1 -> getText(R.string.detail_next_page)
                else -> null
            }
        }
    }
}
